lib/route/br_vlan.c

```diff
@@ -0,0 +1,1027 @@
+		return -NLE_INVAL;
+
+	msg = nlmsg_alloc_simple(RTM_NEWVLAN, NLM_F_REQUEST);
+	if (msg == NULL)
+		return -NLE_NOMEM;
+
+	bvm.ifindex = opts_list->ifindex;
+	bvm.family = AF_BRIDGE;
+
+	err = nlmsg_append(msg, &bvm, sizeof(bvm), NLMSG_ALIGNTO);
+	if (err < 0)
+		goto free_msg;
+
+	nl_list_for_each_entry(opts, &opts_list->gopts_list, list_node) {
+		attr = nla_nest_start(msg, BRIDGE_VLANDB_GLOBAL_OPTIONS);
+		if (!attr) {
+			goto free_msg;
+		}
+
+		if (opts->mask & BR_VLAN_ATTR_GOPTS_ID)
+			NLA_PUT_U16(msg, BRIDGE_VLANDB_GOPTS_ID, opts->o_id);
+		if (opts->mask & BR_VLAN_ATTR_GOPTS_RANGE)
+			NLA_PUT_U16(msg, BRIDGE_VLANDB_GOPTS_RANGE,
+				    opts->o_range);
+		if (opts->mask & BR_VLAN_ATTR_GOPTS_MCAST_SNOOPING)
+			NLA_PUT_U8(msg, BRIDGE_VLANDB_GOPTS_MCAST_SNOOPING,
+				   opts->o_mcast_snooping);
+		if (opts->mask & BR_VLAN_ATTR_GOPTS_MCAST_IGMP_VERSION)
+			NLA_PUT_U8(msg, BRIDGE_VLANDB_GOPTS_MCAST_IGMP_VERSION,
+				   opts->o_mcast_igmp_version);
+		if (opts->mask & BR_VLAN_ATTR_GOPTS_MCAST_MLD_VERSION)
+			NLA_PUT_U8(msg, BRIDGE_VLANDB_GOPTS_MCAST_MLD_VERSION,
+				   opts->o_mcast_mld_version);
+		if (opts->mask & BR_VLAN_ATTR_GOPTS_MCAST_LAST_MEMBER_CNT)
+			NLA_PUT_U32(msg,
+				    BRIDGE_VLANDB_GOPTS_MCAST_LAST_MEMBER_CNT,
+				    opts->o_mcast_last_member_cnt);
+		if (opts->mask & BR_VLAN_ATTR_GOPTS_MCAST_STARTUP_QUERY_CNT)
+			NLA_PUT_U32(msg,
+				    BRIDGE_VLANDB_GOPTS_MCAST_STARTUP_QUERY_CNT,
+				    opts->o_mcast_startup_query_cnt);
+		if (opts->mask & BR_VLAN_ATTR_GOPTS_MCAST_LAST_MEMBER_INTVL)
+			NLA_PUT_U64(msg,
+				    BRIDGE_VLANDB_GOPTS_MCAST_LAST_MEMBER_INTVL,
+				    opts->o_mcast_last_member_intvl);
+		if (opts->mask & BR_VLAN_ATTR_GOPTS_MCAST_MEMBERSHIP_INTVL)
+			NLA_PUT_U64(msg,
+				    BRIDGE_VLANDB_GOPTS_MCAST_MEMBERSHIP_INTVL,
+				    opts->o_mcast_membership_intvl);
+		if (opts->mask & BR_VLAN_ATTR_GOPTS_MCAST_QUERIER_INTVL)
+			NLA_PUT_U64(msg,
+				    BRIDGE_VLANDB_GOPTS_MCAST_QUERIER_INTVL,
+				    opts->o_mcast_querier_intvl);
+		if (opts->mask & BR_VLAN_ATTR_GOPTS_MCAST_QUERY_INTVL)
+			NLA_PUT_U64(msg, BRIDGE_VLANDB_GOPTS_MCAST_QUERY_INTVL,
+				    opts->o_mcast_query_intvl);
+		if (opts->mask & BR_VLAN_ATTR_GOPTS_MCAST_QUERY_RESPONSE_INTVL)
+			NLA_PUT_U64(
+				msg,
+				BRIDGE_VLANDB_GOPTS_MCAST_QUERY_RESPONSE_INTVL,
+				opts->o_mcast_query_response_intvl);
+		if (opts->mask & BR_VLAN_ATTR_GOPTS_MCAST_STARTUP_QUERY_INTVL)
+			NLA_PUT_U64(
+				msg,
+				BRIDGE_VLANDB_GOPTS_MCAST_STARTUP_QUERY_INTVL,
+				opts->o_mcast_startup_query_intvl);
+		if (opts->mask & BR_VLAN_ATTR_GOPTS_MCAST_QUERIER)
+			NLA_PUT_U8(msg, BRIDGE_VLANDB_GOPTS_MCAST_QUERIER,
+				   opts->o_mcast_querier);
+		if (opts->mask & BR_VLAN_ATTR_GOPTS_MSTI)
+			NLA_PUT_U16(msg, BRIDGE_VLANDB_GOPTS_MSTI,
+				    opts->o_msti);
+
+		nla_nest_end(msg, attr);
+	}
+
+	*result = msg;
+	return 0;
+
+nla_put_failure:
+	nla_nest_cancel(msg, attr);
+
+free_msg:
+	nlmsg_free(msg);
```

Comment:
don't use any explicit free/unref/put for variables on the stack (obviously, it's still necessary in functions like `rtnl_br_vlan_gopts_entry_free()`.

In this case, use `_nl_auto_nl_msg struct nl_msg *msg = NULL;` and drop `goto free_msg`.

The `nla_put_failure` label will still be necessary, but it doesn't need to do anything except `return -EMSGSIZE;`. It doesn't need the `nla_nest_cancel()`, because the message is anyway freed.

Also, you will need
```C
    *result = _nl_steal_pointer(&msg);
    return 0;
```
to explicitly transfer ownership.